### ISSUE TITLE
cast discards const in safestack.h

### ISSUE THREAD

**Issue by msalle** [openssl/openssl#1384]
In 5 places in safestack.h, all in the definition of SKM_DEFINE_STACK_OF(t1, t2, t3) (which is used for the definitions of e.g. the DEFINE_STACK_OF() macros which in turn are used in many header files) a cast from a const pointer to a non-const pointer is done:

- openssl/safestack.h:62
- openssl/safestack.h:66
- openssl/safestack.h:82
- openssl/safestack.h:86
- openssl/safestack.h:112

Compile as gcc -Wcast-qual test.c the following:

`#include <openssl/safestack.h>`
`int main(void)  { return 0; }`

I'm pretty sure it's a regression introduced in commit 411abf2dd37974a5baa54859c1abcd287b3c1181.

**Comment by richsalz** [comment]
Are you sure you are looking at current master?  What compiler and version?  When I try your sample program it won't compile (needs e_os2.h) and then I get other warnings.

Are you just reading the code?  Can you show a small complete program that compiles?

Of the lines you listed, only the "deep_copy" function seems to have an incorrect cast, missing a const.

**Comment by msalle** [comment]
Yes, I'm sure. I'm using either gcc (SUSE Linux) 4.8.5 or clang version 3.7.0 (tags/RELEASE_370/final 246586).
Current master (i.e. tree at commit 790555d6756285b3ec18e3efbb195cf33f217d8f) build with same gcc:
./Configure --prefix=/scratch/openssl-master linux-x86_64
make -j && make -j install
Then
gcc -Wcast-qual -I/scratch/openssl-master/include test.c 2> warnings_gcc.txt
attached is the output of stderr for both compilers and test.c itself in one zip file.
[test_program.zip](https://github.com/openssl/openssl/files/400988/test_program.zip)

**Comment by richsalz** [comment]
Ah, okay, now I see.  It's not the first argument, it's the second.  Let me take a look.

**Comment by richsalz** [comment]
(okay sometimes it's the first)

**Comment by richsalz** [comment]
can you try https://github.com/openssl/openssl/pull/1397 and see what you get?

**Comment by msalle** [comment]
I still get some warnings. It's the `const` in the second argument of the definitions of
- `DEFINE_SPECIAL_STACK_OF_CONST` (line 123/124)
- `DEFINE_STACK_OF_CONST` (line 125)

i.e. calling the `SKM_DEFINE_STACK_OF()` macro with a const for the second argument. The other warning is due to the function definition of
- `*sk_##t1##_deep_copy` (line 108)

where the argument `const STACK_OF(t1) *sk` is cast to `(OPENSSL_STACK *)` i.e. to (`struct stack_st *`)

**Comment by richsalz** [comment]
Pushed some fixes.  Please post the compiler output again if you can try this; thanks.

**Comment by msalle** [comment]
Sure, hereby the gcc output from gcc with -Wcast-qual
[warning_1397_gcc.txt](https://github.com/openssl/openssl/files/403936/warning_1397_gcc.txt)

**Comment by richsalz** [comment]
Okay, I think I have it fixed now :)  Please try agai.

**Comment by msalle** [comment]
Great, I can confirm the warnings are gone!

**Comment by richsalz** [comment]
Thanks for all your help!  Ping @levitte, for const-ification review? ☺

**Comment by mattcaswell** [comment]
Can this be closed now?

**Comment by richsalz** [comment]
Once the PR is closed, yes.  That one needs a second reviewer.
